### ISSUE TITLE
Terraform building can be done in parallel instead of serially

### ISSUE THREAD

**Issue by Shelda-On** [freeorion/freeorion#2676]
### Environment
* **FreeOrion Version**: 0.4.8
* **Operating System**:  MacOS High Sierra
* **Graphic card used**: NA
* **Fetched as** 0.4.8 Stable build

### Description
In v0.4.8 I find that I can stack Terraforming. For example Egassem have good on inferno. If I have a Desert I want to terraform. I have to terraform it 4 times. Each time takes 8 turns to build + 1 to complete. So this will be a total of 45 turns. and each of the terraforms becomes more expensive (2nd more than first, third more than second, forth most of all.) I find that I can start doing 4 terraforms in parallel. They all cost the same amount as the first one. Also 7 turns are in parallel, followed by 1 turn to build the last and one to complete it. I think this comes to a total of 11 turns (might be 14) and only the last turn of each is more expensive.

Note:  this also applies to the Terraforming Reversion.

### Expected Result
 think I found the issue. The files FreeOrion.app/Contents/Resources/default/scripting/buildings/TERRAFORM.focs.txt (and I assume the TERRAFORM_REVERT.focs.txt would need a similar change.)

````
EnqueueLocation = And [
Planet
OwnedBy empire = Source.Owner
Not Contains Building name = "BLD_TERRAFORM_REVERT"
Not Contains Building name = "BLD_GAIA_TRANS"
Not Contains Building name = "BLD_TERRAFORM" # Line added
Not HasSpecial name = "GAIA_SPECIAL"
Not Enqueued type = Building name = "BLD_TERRAFORM_REVERT"
Not Enqueued type = Building name = "BLD_GAIA_TRANS"
Not Enqueued type = Building name = "BLD_TERRAFORM" # Line added
]
````

**Comment by Shelda-On** [comment]
See discussion in https://www.freeorion.org/forum/viewtopic.php?f=25&t=11443&p=98669#p98669

**Comment by Shelda-On** [comment]
I tested the changes for Terraforming but not the changes for Revert Terraforming.

**Comment by Shelda-On** [comment]
I can't seem to create a branch to address this.  so someone else will have to clear it.

**Comment by geoffthemedio** [comment]
Is this still an issue with a recent test release / 0.4.9 pre-release build?

**Comment by MatGB** [comment]
Not a bug. Feature, you're meant to be able to stack them until someone can
redo terraforming to set target environments or use a focus effect.

If there is a bug, the cost multiplier on completion of a building might
not be working, but that's not the issue described. Can't test, moving
house, computer in storage

Terraforming projects are meant to be stacked, it's not meant to take
most of a game to micromanage a single world

-- 
Mat Bowles

**Comment by agrrr3** [comment]
@MatGB Oh,hi mat!!! :) You beat me to this.

If there is a bug, it is not the enqueability. 

Adding `Not Enqueued type = Building name = "BLD_TERRAFORM"` to the location condition (not the enqueueLocation) would still enable to enqueue, but building would be serial which should also fix the cost multiplier.

**Comment by Oberlus** [comment]
A sight for sore eyes!

I agree with agrrr3: being able to enqueue several at one but not to invest PPs in more than one is a nice fix, as long as cost of queued extra terraformings is updated after each one is finished.

**Comment by Shelda-On** [comment]
So is the intention to allow 2, 3 or 4 terraforming to be built in parallel with different PPs?  Or to enqueue 2, 3 or 4 of them but have them be produced one at a time without having to put new ones in?  Or something like "when a terraforming focus is selected" or "a build is created" then keep doing things that make the planet better (terraform and gaiaform) until it is done?  (If this is a focus in place of research or industry, I would probably rather manually control the terraforming additions so that I can get benefits of both at the same time.)

**Comment by Oberlus** [comment]
> So is the intention to [...] enqueue 2, 3 or 4 of them but have them be produced one at a time without having to put new ones in?

This, yes.
The idea is to be able to do all the clicks at the same turn (enqueue all buildings at once when colonised) and not having to revisit the colony several times, without breaking the intended mechanic of further-terraforming-becomes-harder.

**Comment by MatGB** [comment]
When I last did something with this script,  the intent was to allow you to
decide how you wanted the planet to end up at one time, set it up to do
that, and leave it, not needing to go back and do more work repeatedly

At the time, we didn't have the X% done thing in the costs multiplier, so
unfinished buildings would be partially complete and need more work, the
changes do make this a little unbalanced I guess but it's an edge case

I dislike terraforming by multiple buildings anyway, it's always been a
stopgap, as you've got way more contributors now, perhaps brainstorm a
better way, perhaps focus based, maybe even with remote targetting so you
can prepare an outpost?

(Life has been weird ,I hope to start playing again at some point ,I never
unsubbed from github and that allows reply from email—I know this topic
backwards so thought a comment worth it )

-- 
Mat Bowles
ƃuᴉɯɐǝɹp sʇᴉɐʍ nɥlnɥʇƆ pɐǝp ɥǝʎl,ɹ ʇɐ ǝsnoH sᴉH uI

**Comment by NathanZook** [comment]
The naïve approach would be to have each terraform level be different internally.  So TERRAFORM_4 depends on TERRAFORM_3 ...  However, to the player, there is just one terraform building that can be built multiple times.  (In the queue, the difference should probably be visible.)
Furthermore, if the current species is, say, two steps away, then they cannot queue up TERRAFORM_3.

Reverting terraforming works in the obvious fashion.

**Comment by Oberlus** [comment]
Another way could be to make terraforming a unique building (can't enqueue more than once) that upon completion can self-enqueue itself as long as current environment is not Good. The player can remove it from the queue before that, being warned by the sitrep each time a phase is completed.

**Comment by agrrr3** [comment]
This issue can be closed as terraforming got completely reimplemented. If there are bugs with the current implementation, we need a new issue

**Comment by Vezzra** [comment]
Setting this to the 0.5 milestone, assuming the reimplemented terraforming is part of the upcoming release. If I'm mistaken, please change the milestone to post release.

**Comment by o01eg** [comment]
It's correct. Terraforming was reimplemented here https://github.com/freeorion/freeorion/pull/3634
